### PR TITLE
Cache ropes for Regexp#to_s

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

File: src/main/java/org/truffleruby/core/regexp/RegexpNodes.java [hunk 1]
```diff
@@ -693,13 +693,28 @@ public DynamicObject source(DynamicObject regexp) {
...
-        @TruffleBoundary
-        @Specialization
+        @Specialization(guards = "hasString(regexp)")
+        public DynamicObject toSFast(DynamicObject regexp) {
+            return createString(Layouts.REGEXP.getString(regexp));
+        }
+
+        @Specialization(guards = "!hasString(regexp)")
         public DynamicObject toS(DynamicObject regexp) {
+            Rope byteList = createRope(regexp);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/regexp/RegexpNodes.java]: This should be named `rope`

Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/regexp/RegexpNodes.java]: In keeping with the existing code style, it should be `final` as well.

File: src/main/java/org/truffleruby/core/regexp/RegexpNodes.java [hunk 2]
```diff
@@ -693,13 +693,28 @@ public DynamicObject source(DynamicObject regexp) {
...
+            Layouts.REGEXP.setString(regexp, byteList);
+            return createString(byteList);
+        }
+
+        @TruffleBoundary
+        protected Rope createRope(DynamicObject regexp) {
             final ClassicRegexp classicRegexp = ClassicRegexp.newRegexp(getContext(), Layouts.REGEXP.getSource(regexp), Layouts.REGEXP.getRegex(regexp).getOptions());
-            return createString(classicRegexp.toByteList());
+            RopeBuilder byteList = classicRegexp.toByteList();
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/core/regexp/RegexpNodes.java]: This should be named `ropeBuilder` or just `builder`. It's unfortunate that some variable names weren't caught in a previous refactor.

**Comment by aardvark179** [comment]
I did try going for a `@Cached` solution, but found we were exceeding the cache size too often. It's true that `Regexp#to_s` is rarely called explicitly, but it is frequently called as part of making interpolated regexp literals and unions. Unions were the main area where using `@Cached` failed, so I might look at whether we could avoid the `to_s` calls entirely most of the time.

Many of the lexers and other utils used by Jekyll use interpolated regexp literals or construct unions, and they do so more often than you'd think.

**Comment by eregon** [comment]
> Many of the lexers and other utils used by Jekyll use interpolated regexp literals or construct unions, and they do so more often than you'd think.

But do they do that on the fast-path? Regex.union or interpolated Regexp are going to be order of magnitudes more expensive than copying a byte array of the Regexp source, no? Or is there hidden complexity in Regexp#to_s?

**Comment by aardvark179** [comment]
Sadly they do it enough on the fast path for this stuff to be important. I've taken another look at unions, and managed to do things slightly more cleverly there to avoid calling `to_s` on that fast path, so I changed this to cache on the site rather than the regexp.

File: src/main/java/org/truffleruby/core/regexp/RegexpNodes.java [hunk 3]
```diff
@@ -693,11 +693,24 @@ public DynamicObject source(DynamicObject regexp) {
...
+            final Rope rope = createRope(regexp);
+            return createString(rope);
+        }
+
+        @TruffleBoundary
+        protected Rope createRope(DynamicObject regexp) {
             final ClassicRegexp classicRegexp = ClassicRegexp.newRegexp(getContext(), Layouts.REGEXP.getSource(regexp), Layouts.REGEXP.getRegex(regexp).getOptions());
-            return createString(classicRegexp.toByteList());
+            final Rope rope = classicRegexp.toByteList().toRope();
+            return rope;
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/regexp/RegexpNodes.java]: This can just return without assigning.

**Comment by eregon** [comment]
This looks good to me.
We should squash those commits as the final diff is tiny compared to the intermediate diffs.

**Comment by eregon** [comment]
Re calling Regexp#to_s (or Regexp#source) for Regexp.union, MRI doesn't do it so we should be fine:
```ruby
[1] pry(main)> r=/a/
=> /a/
[2] pry(main)> def r.to_s; "foo" end
=> :to_s
[3] pry(main)> r
=> /a/
[4] pry(main)> r.to_s
=> "foo"
[5] pry(main)> Regexp.union(r)
=> /a/
[6] pry(main)> Regexp.union(r, r)
=> /(?-mix:a)|(?-mix:a)/
```

**Pull request merged**